### PR TITLE
Bug fix: Multiple abbreviations & incorrect casing

### LINKED ISSUE #125: Multiple abbreviations not handled correctly
## Overview
In working on #119, I noticed that [the code for handling abbreviations](https://github.com/monarch-initiative/omim/blob/main/omim2obo/main.py#L184-L215) could use a little work. There are not many cases of multiple symbols / abbreviations. In fact, I've just found 2 cases:

[OMIM:126370](https://omim.org/entry/126370): DNA, SATELLITE, III; HS3; D1Z1
[OMIM:171820](https://omim.org/entry/171820): PHOSPHATASE, SALIVARY ACID, A; SACP; ACPS

I'm not sure how many issues are being caused by improper handling of multiple abbreviations, but there appears to be at least some...

## Issue 1: `rdfs:label`
It adds the `rdfs:label` for one of the abbreviations. And apparently, given the way the code works (i.e. setting abbrev as primary label for genes), this only affects `OMIM:171820`.

## Issue 2: Synonyms & "modified included label"
There should be an outer `for` loop here over list of abbreviations. There are these 3 triples for synonyms, and another one for a "modified included label" (not sure why we're creating this).
https://github.com/monarch-initiative/omim/blob/dc3c79a5606a495cd7a08623ed4ac17c234d0575/omim2obo/main.py#L207-L215

I don't think there should be `abbrev` and `abbr`. There should just be `abbreviations`.

## Questions
- [ ] 1. Am I correct in thinking that the two examples above are really 2 different symbols, and not 1?  _(I think so)_
  - I think so, but maybe the actual symbols are "HS3; D1Z1" and "SACP; ACPS", and `;` are simply allowed within a symbol? I say this because the header of `mimTitles.txt` includes fields such as `Preferred Title; symbol`, `Alternative Title(s); symbol(s)`, `Included Title(s); symbols`. This indicates that multiple symbols can happen for alternative or included titles, but not preferred titles. However, I think this is an oversight, and the field name should be "`Preferred Title; symbol(s)`".

### LINKED ISSUE #126: Bug: Symbols not always uppercase
## Overview
In working on #119, I noticed that some of the the `mondo#abbreviation` we're adding (these are actually _symbols_) are lowercase.

This is not expected. `mondo#abbreviation`s should all be uppercase.
And also I would expect that symbols always be uppercase too, as I learned from Trish. And a [cursory search](https://en.wikipedia.org/wiki/Gene_nomenclature#Human) seems to corroborate that.

The source of this bug in the code appears to be [`cleanup_label()`](https://github.com/monarch-initiative/omim/blob/dc3c79a5606a495cd7a08623ed4ac17c234d0575/omim2obo/parsers/omim_entry_parser.py#L168) and [`_detect_abbreviations()`](https://github.com/monarch-initiative/omim/blob/dc3c79a5606a495cd7a08623ed4ac17c234d0575/omim2obo/parsers/omim_entry_parser.py#L125).

I think it could be that (a) there's bugs in these functions, or (b) these functions simply should not be applied to symbols--at least not the ones on "Preferred title; symbol", and are only supposed to be used on other labels or parts of labels.

## Examples
[OMIM:126370](https://omim.org/entry/126370) - _hs3_
```ttl
OMIM:126370 a owl:Class ;
    rdfs:label "dna, satellite, 3" ;
    MONDO:exclusionReason MONDO:excludeTrait ;
    oboInOwl:hasExactSynonym "D1Z1",
        "dna, satellite, 3",
        "hs3" .
```

## Sub-tasks
- [ ] 1. Fix any related bugs
  - [ ] Could this bug be because the parser is expecting only 1 symbol, but this one has 2 symbols?
- [ ] 2. Refactor?: Should we just capitalize all symbols, rather than change casing based on logic in `cleanup_label()`?

### PR DESCRIPTION
**Pull request by joeflack4**
resolves #125 
resolves #126

## Changes
Bug fix: Multiple abbreviations & incorrect casing
- When multiple preferred symbols, they were not being handled correctly. This also resulted in some of these symbols being added with incorrect casing.

### CODE REVIEW

File: omim2obo/main.py [hunk 1]
```diff
@@ -197,28 +199,27 @@ def omim2obo(use_cache: bool = False):
         else:
             pass
 
-        if use_abbrev_over_label and abbrev:
-            graph.add((omim_uri, RDFS.label, Literal(abbrev)))
+        if use_abbrev_over_label and pref_symbols:
+            if len(pref_symbols) > 1:
```
Comment (twhetzel) [inline omim2obo/main.py:203]: There is a section that describes the file format for "Entry Data" on https://omim.org/help/api#1_1. It describes the "preferredTitle" as "title and symbol, delimited with ';'". I analyzed the mimTitles.txt file from 4-Sep-2024 and there are 2 entries where there is >1 symbol in this field.

These are: 

MIM Number | Preferred Title; symbol
-- | --
126370 | DNA, SATELLITE, III; HS3; D1Z1
171820 | PHOSPHATASE, SALIVARY ACID, A; SACP; ACPS

Therefore, for the "preferredTitle" field, anything after the first `;` can be considered a symbol.

Keep in mind "includedTitles" are formatted differently so what is decided for the "preferredTitle" field should not be confused with what needs to happen for values in the "includedTitles" field.

Comment (joeflack4) [reply omim2obo/main.py:203]: That corroborates my reading of the docs and analysis as well (described a little in #125).

> Therefore, for the "preferredTitle" field, anything after the first ; can be considered a symbol.

I believe the code does that correctly now!

> Keep in mind "includedTitles" are formatted differently

Yeah, I wasn't sure. That issue (#118), I'll either do in the greater #128 PR or a separate PR.

Comment (twhetzel) [reply omim2obo/main.py:203]: The script should not fail when running if >1 abbreviation is found in the "preferredTitle" field.

Comment (joeflack4) [reply omim2obo/main.py:203]: Ah. Yes, that is my current failsafe until we decide what to do about these cases.

It doesn't just fail if there is more than 1 abbreviation. This is a more complex case.

I've only set it to fail if we are using the symbol/abbreviation as the main `rdfs:label` rather than a Synonym.

The reason for that is because if we want to use an symbol as a label, I don't know yet which symbol to pick if there are >1, or whether I should just include all of them as a concatenated string.

The only situation where we do this; set the abbrev as a label, is if `omim_type == OmimType.GENE or omim_type == OmimType.HAS_AFFECTED_FEATURE:  # * or +`.

Note that there are no current cases where this is failing in the code.

Comment (twhetzel) [reply omim2obo/main.py:203]: Thanks for the explanation, since this check is only when `omim_type == OmimType.GENE or omim_type == OmimType.HAS_AFFECTED_FEATURE:  # * or +` is in place then it's ok to leave this.

File: omim2obo/main.py [hunk 2]
```diff
@@ -197,28 +199,27 @@ def omim2obo(use_cache: bool = False):
...
+            graph.add((omim_uri, oboInOwl.hasExactSynonym, Literal(abbreviation)))
+            axiom = BNode()
+            graph.add((axiom, RDF.type, OWL.Axiom))
+            graph.add((axiom, OWL.annotatedSource, omim_uri))
+            graph.add((axiom, OWL.annotatedProperty, oboInOwl.hasExactSynonym))
+            graph.add((axiom, OWL.annotatedTarget, Literal(abbreviation)))
+            graph.add((axiom, oboInOwl.hasExactSynonym, MONDONS.abbreviation))
+        # todo: .clean() should accept all symbols/abbrevs (preferred, alt, included), else [] instead of str/None
+        #  - See: https://github.com/monarch-initiative/omim/issues/129
+        abbrev: Union[str, None] = [None] if not pref_symbols else pref_symbols[0]
```
Comment (twhetzel) [inline omim2obo/main.py:220]: What is this line here doing?

File: omim2obo/main.py [hunk 3]
```diff
@@ -197,28 +199,27 @@ def omim2obo(use_cache: bool = False):
...
+        for alt_label in other_labels:
+            graph.add((omim_uri, oboInOwl.hasExactSynonym, Literal(label_cleaner.clean(alt_label, abbrev))))
```
Comment (twhetzel) [inline omim2obo/main.py:222]: To check, this cleaning of an abbreviation from an Alternative label will be handled in a separate PR?

**Comment by twhetzel** [comment]
@joeflack4 if this is ready for PR review, can you change it out of Draft mode?

File: omim2obo/main.py (hunk 3 above)
Comment (joeflack4) [reply omim2obo/main.py:222]: Yes, not this PR. That applies to this one:
- #129

File: omim2obo/main.py (hunk 2 above)
Comment (joeflack4) [reply omim2obo/main.py:220]: `[None]` is not correct and I changed to `None`.

Basically, `abbrev` gets passed to `.clean_label()`, and until I tackle #129, it doesn't accept an empty list for that arg; it expects `None`. So if there are no symbols I convert to `None` and pass that instead.

File: omim2obo/main.py (hunk 1 above)
Comment (joeflack4) [reply omim2obo/main.py:203]: @twhetzel I'm surprised! `mimTitles.txt` was just updated, and a MIM is now triggering this error!

[OMIM:171820](https://omim.org/entry/171820): _PHOSPHATASE, SALIVARY ACID, A; SACP; ACPS_

I made these changes:
* Warning now instead of error
* Picks the first symbol to use as label

I don't know how concerned we have to be about this, because these gene classes don't really get into Mondo anyway.

File: omim2obo/main.py [hunk 4]
```diff
@@ -164,63 +164,69 @@ def omim2obo(use_cache: bool = False):
...
-                graph.add((axiom, OWL.annotatedProperty, oboInOwl.hasExactSynonym))
-                graph.add((axiom, OWL.annotatedTarget, Literal(abbr)))
-                graph.add((axiom, oboInOwl.hasExactSynonym, MONDONS.abbreviation))
-        for exact_label in exact_labels:
-            graph.add((omim_uri, oboInOwl.hasExactSynonym, Literal(label_cleaner.clean(exact_label, abbrev))))
-        for label in other_labels:
-            graph.add((omim_uri, oboInOwl.hasExactSynonym, Literal(label_cleaner.clean(label, abbrev))))
+            graph.add((omim_uri, RDFS.label, Literal(label_cleaner.clean(pref_title))))
+
+        # todo: .clean() should accept all symbols/abbrevs (preferred, alt, included), else [] instead of str/None
```
Comment (twhetzel) [inline omim2obo/main.py:206]: I don't think `.clean()` ever needs to be used on symbols/abbrevs since no case changes need to be made to these values.

Is this a reference to the possible future work about symbols/abbrevs being used to inform the casing of symbols found in preferred Titles as you pointed out in our discussion? If yes, then can you make the comment more clear to indicate that?

Comment (twhetzel) [reply omim2obo/main.py:206]: Ok, I think that comment is for issue 129. If it said something like `.clean()` should accept a list of abbreviations vs. only one abbreviation value then that would be more clear.

File: omim2obo/main.py [hunk 5]
```diff
@@ -164,63 +164,69 @@ def omim2obo(use_cache: bool = False):
...
+        #  - See: https://github.com/monarch-initiative/omim/issues/129
+        abbrev: Union[str, None] = None if not pref_symbols else pref_symbols[0]
+
+        # Add synonyms
+        graph.add((omim_uri, oboInOwl.hasExactSynonym, Literal(label_cleaner.clean(pref_title, abbrev))))
```
Comment (twhetzel) [inline omim2obo/main.py:211]: This line looks new since the [last commit we reviewed](https://github.com/monarch-initiative/omim/pull/130/commits/d41df383cee4656c176b0905e2650185319b263b). Do you have the diff of the omim.owl files or the .ttl file that is created with this newer version of the code?

**Review by twhetzel — COMMENTED** [review]
Overall, looks ok. I would like to see the new .ttl file that is generated from these changes since Line 211 looks different from the code version we last reviewed together.

**Review by twhetzel — CHANGES_REQUESTED** [review]
Adding another review so the status can be Request changes.

File: omim2obo/main.py [hunk 6]
```diff
@@ -164,63 +164,69 @@ def omim2obo(use_cache: bool = False):
...
+            graph.add((axiom, OWL.annotatedProperty, oboInOwl.hasExactSynonym))
+            graph.add((axiom, OWL.annotatedTarget, Literal(abbreviation)))
+            graph.add((axiom, OBOINOWL.hasSynonymType, MONDONS.abbreviation))
+
+        # Add 'included' entry properties
         for included_label in cleaned_inc_labels:
             graph.add((omim_uri, URIRef(INCLUDED_URI), Literal(label_cleaner.clean(included_label, abbrev))))
+        if label_endswith_included_alt or label_endswith_included_inc:
+            graph.add(
+                (omim_uri, RDFS['comment'], Literal("This term has one or more labels that end with ', INCLUDED'.")))
```
Comment (twhetzel) [inline omim2obo/main.py:229]: Rather than the string, you can still use `included_detected_comment`.

File: omim2obo/main.py [hunk 7]
```diff
@@ -164,63 +164,69 @@ def omim2obo(use_cache: bool = False):
...
+            graph.add((axiom, RDF.type, OWL.Axiom))
+            graph.add((axiom, OWL.annotatedSource, omim_uri))
+            graph.add((axiom, OWL.annotatedProperty, oboInOwl.hasExactSynonym))
+            graph.add((axiom, OWL.annotatedTarget, Literal(abbreviation)))
+            graph.add((axiom, OBOINOWL.hasSynonymType, MONDONS.abbreviation))
+
+        # Add 'included' entry properties
         for included_label in cleaned_inc_labels:
             graph.add((omim_uri, URIRef(INCLUDED_URI), Literal(label_cleaner.clean(included_label, abbrev))))
+        if label_endswith_included_alt or label_endswith_included_inc:
```
Comment (twhetzel) [inline omim2obo/main.py:227]: Let's check that moving this line here does not change which entries have the `included_detected_comment` comment.

File: omim2obo/main.py (hunk 4 above)
Comment (joeflack4) [reply omim2obo/main.py:206]: Yes, you're correct. But it is actually not just accepting a list of abbreviations. It is that it should pass all abbreviations of any of preferred, alt, or included. Currently, it is only passing the 1st preferred abbreviation.

I updated the comment to:
> todo: .clean()/.cleanup_label() 2nd param `explicit_abbrev` should be List[str] instead of str. And below, should pass all symbols/abbrevs from each of preferred, alt, included each time it is called. If no symbols for given term, should pass empty list. See: https://github.com/monarch-initiative/omim/issues/129

File: omim2obo/main.py (hunk 5 above)
Comment (joeflack4) [reply omim2obo/main.py:211]: Very observant. This line was there when I ran it with you yesterday and looked at the diff. Pretty sure it had been pushed by that time, but probably not when you reviewed. If it was missing when you reviewed, that was a mistake; it should've never have been missing.

This operation isn't new per se. It is a refactor of:

https://github.com/monarch-initiative/omim/blob/db4ddd2e88a773d1c5743c82529c665edc1a925c/omim2obo/main.py#L210-L211

Previously, the code had referred to "preferred title" as "exact labels". IDK why. It also assumed there could be more than 1 preferred title, which there can't. So IIRC, this is also part of what was causing lowercasing in one of the symbols for [those 2 symbol cases](https://github.com/monarch-initiative/omim/issues/125)

Comment (joeflack4) [reply omim2obo/main.py:211]: Also FYI, this 1 line is the topic of this issue as well:
- #131

Comment (joeflack4) [reply omim2obo/main.py:212]: > would like to see the new .ttl file

I undeleted the [google drive folder](https://drive.google.com/drive/folders/1yeQAfh_yzI8vCEzkaBblF6-B__EEJFsT) w/ the uploads we were looking yesterday, reran and created a new `omim.ttl`, and uploaded it there (renamed to `after3.ttl`).

I looked at the `md5`s, but indeed as we were discussing yesterday, it is non-deterministic, and even when running it back-to-back, I got a different hash each time. Comparing `.ofn` would be more ideal, but it looks like I deleted the one we were looking at yesterday (directly from my IDE, so it's not in my trash either).

File: omim2obo/main.py (hunk 6 above)
Comment (joeflack4) [reply omim2obo/main.py:230]: Oh, alright. Is that what you added? I thought it was something I added. Normally when I add a separate line for a comment, it has to do with indentation and trying to minimize the number of lines; this refactor yielded the same results in less lines. But I will change it back.

File: omim2obo/main.py (hunk 7 above)
Comment (joeflack4) [reply omim2obo/main.py:228]: As with [this thread](https://github.com/monarch-initiative/omim/pull/130#discussion_r1748276110), I'm pretty sure this change was already there when we looked at the `.ofn` diffs, etc, on the call yesterday, so all should be well. But I did [run and upload another output](https://github.com/monarch-initiative/omim/pull/130#discussion_r1748972938), so feel free to check!

Comment (joeflack4) [reply omim2obo/main.py:228]: FYI I think these additional comments, along w/ some re-ordering, make the code much more readable!

File: omim2obo/main.py (hunk 5 above)
Comment (twhetzel) [reply omim2obo/main.py:212]: Thanks for the updated file, the abbreviations look good.

**Review by twhetzel — APPROVED** [review]
lgtm
